ProjFS.Mac/PrjFSKext/PrjFSKext/Memory.hpp

```diff
@@ -5,8 +5,18 @@ kern_return_t Memory_Init();
 kern_return_t Memory_Cleanup();
 
 void* Memory_Alloc(uint32_t size);
+// Variant which is safe to call while lock is held, but with low chance of success.
```

Comment:
> Which lock?

Any lock that may impede progress of a thread that may be preventing the allocation to complete.

> Under what circumstances would allocating memory block?

If the allocator must itself acquire a lock in order to fulfil the request. So there may be a lock free fast path if there is a free block of memory in the suitably sized zone, but if it triggers an allocation of fresh pages in the zone, that probably requires a lock. That in itself wouldn’t necessarily be bad, but in low memory situations, this can lead to a deadlock as hinted at above. Essentially, if there are no free pages available, some memory may need to be pages out to disk. This causes I/O, and our callbacks are technically in the path of such I/O. In theory, the auth checks should already have passed by the time a memory mapped file gets flushed to disk, but I wouldn’t be surprised if there were specific circumstances where our callback is called as a result of a pageout. If that callback then attempts to acquire the lock we originally held while trying to allocate memory, you’ve got yourself a deadlock.

I’ve not seen this sequence of events happen in kauth vnode listeners, but I’d still be wary. I’ve definitely encountered this type of deadlock in block device driver kexts and MAC framework callbacks (these tend to be a little higher traffic than kauth) that weren’t careful about allocations. I think we’re a little more at risk of this than normal because we have this situation where the kext blocks progress until it hears back from user space.

To prevent this kind of situation entirely, you’d probably need to ensure you never have to allow on a critical path at all, but allocating without holding a lock should be safe enough - I’ve never seen a report where the problem still occurred in that case.

Reply:
@pmj thanks for all the details and helping me get up to speed on this.  A few more questions regarding the memory allocation:

> Any lock that may impede progress of a thread that may be preventing the allocation to complete.

When you say "any lock" do you mean any lock that is owned\managed by the allocator?  Or is it any lock used in kernel space (e.g. would our `s_virtualizationRootsLock` count because it uses kernel locking primitives)?

> This causes I/O, and our callbacks are technically in the path of such I/O. In theory, the auth checks should already have passed by the time a memory mapped file gets flushed to disk, but I wouldn’t be surprised if there were specific circumstances where our callback is called as a result of a pageout. If that callback then attempts to acquire the lock we originally held while trying to allocate memory, you’ve got yourself a deadlock.

Thanks I think I've got it.  Am I correct in summarizing the scenario as the following sequence of events?

1. The allocator acquired a lock
2. The allocator decided to page out to disk
3. Our kext was called (due to paging out to disk)
4. Our kext attempts to allocate memory
5. Deadlock

> To prevent this kind of situation entirely, you’d probably need to ensure you never have to allow on a critical path at all, but allocating without holding a lock should be safe enough - I’ve never seen a report where the problem still occurred in that case.

When you say "where the problem still occurred in that case", do you mean the case of allocating without holding a lock? (i.e. calling `OSMalloc_noblock`?) . Or was this referring to a different case?

Reply:
> When you say "any lock" do you mean any lock that is owned\managed by the allocator? Or is it any lock used in kernel space (e.g. would our `s_virtualizationRootsLock` count because it uses kernel locking primitives)?

Yes, I do mean any lock, and `s_virtualizationRootsLock` counts.

> 1. The allocator acquired a lock
> 2. The allocator decided to page out to disk
> 3. Our kext was called (due to paging out to disk)
> 4. Our kext attempts to allocate memory
> 5. Deadlock

No, that's not quite it; the allocator should be (and as far as I'm aware, *is*) written to not to be vulnerable to quite such a severe weakness. I'm referring to:

1. Our kext acquires `s_virtualizationRootsLock` and finds it needs to resize the mount points array.
2. We allocate a new array while `s_virtualizationRootsLock` is exclusively locked.
3. The allocator doesn't have the memory available without blocking and needs to kick off some paging operations.
4. The same thread or another thread pages out some memory, causing a file write.
5. The file write causes a kauth callback into our kext.
6. Our kext needs to look up the file, so tries to acquire `s_virtualizationRootsLock`, which is already held by our original callback call.
7. Deadlock.

Reply:
> When you say "where the problem still occurred in that case", do you mean the case of allocating without holding a lock? (i.e. calling `OSMalloc_noblock`?) . Or was this referring to a different case?

Not holding any lock while making a blocking allocation, or using `OSMalloc_noblock` to allocate while holding a lock. Note that the latter frequently fails, so you need a backup plan. This generally means you need to drop your lock, use blocking alloc, re-lock and retry your transaction.

Reply:
> I'm glad you called this out is it's something I'll need to account for in the vnode cache implementation.

Yep, definitely. Our virtualisation root array is currently also vulnerable to this. Like this one, it'll be resized extremely rarely, so the chances of the problem occurring are vanishingly small, but I'm planning to fix it. I was waiting with that until we have some generic data structures that solve the problem centrally (like this one) though.